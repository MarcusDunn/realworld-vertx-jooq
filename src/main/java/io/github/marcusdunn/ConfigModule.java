package io.github.marcusdunn;

import dagger.Module;
import dagger.Provides;
import org.apache.commons.configuration2.Configuration;
import org.apache.commons.configuration2.EnvironmentConfiguration;

@Module
public class ConfigModule {
    @Provides
    static Configuration configuration() {
        return new EnvironmentConfiguration();
    }
}
