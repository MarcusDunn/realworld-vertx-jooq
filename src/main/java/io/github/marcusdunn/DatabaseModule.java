package io.github.marcusdunn;

import dagger.Module;
import dagger.Provides;
import io.r2dbc.spi.ConnectionFactories;
import io.r2dbc.spi.ConnectionFactory;
import io.r2dbc.spi.ConnectionFactoryOptions;

@Module
public class DatabaseModule {
    @Provides
    ConnectionFactory connectionFactory() {
        return ConnectionFactories.get(ConnectionFactoryOptions
                .parse("r2dbc:postgresql://localhost:5432/realworld")
                .mutate()
                .option(ConnectionFactoryOptions.USER, "realworld")
                .option(ConnectionFactoryOptions.PASSWORD, "realworld")
                .build()
        );
    }
}
