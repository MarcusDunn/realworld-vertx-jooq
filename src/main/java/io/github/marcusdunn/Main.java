package io.github.marcusdunn;

import dagger.Component;
import io.github.marcusdunn.user.UserModule;
import io.vertx.core.Future;
import io.vertx.core.Vertx;
import io.vertx.core.http.HttpServer;
import io.vertx.core.http.HttpServerOptions;
import io.vertx.ext.web.handler.AuthenticationHandler;
import io.vertx.ext.web.handler.BodyHandler;
import io.vertx.ext.web.openapi.RouterBuilder;
import io.vertx.ext.web.openapi.RouterBuilderOptions;
import java.util.Set;
import java.util.logging.Level;
import java.util.logging.Logger;
import javax.inject.Inject;
import javax.inject.Singleton;

public class Main {
    private final static Logger logger = Logger.getLogger(Main.class.getName());
    private final Vertx vertx;
    private final Set<OperationHandler> operationHandlers;
    private final BodyHandler bodyHandler;
    private final AuthenticationHandler authenticationHandler;
    private final HttpServerOptions httpServerOptions;
    private final RouterBuilderOptions routerBuilderOptions;

    @Inject
    public Main(
            Vertx vertx,
            Set<OperationHandler> operationHandlers,
            BodyHandler bodyHandler,
            AuthenticationHandler authenticationHandler,
            HttpServerOptions httpServerOptions,
            RouterBuilderOptions routerBuilderOptions
    ) {
        this.vertx = vertx;
        this.operationHandlers = operationHandlers;
        this.bodyHandler = bodyHandler;
        this.authenticationHandler = authenticationHandler;
        this.httpServerOptions = httpServerOptions;
        this.routerBuilderOptions = routerBuilderOptions;
    }

    public static void main(String[] args) {
        final var startMillis = System.currentTimeMillis();
        DaggerMain_RealWorld
                .create()
                .main()
                .run()
                .onSuccess(httpServer ->
                        logger.info(() -> "started in " + (System.currentTimeMillis() - startMillis) + "ms")
                );
    }

    public Future<HttpServer> run() {
        return RouterBuilder
                .create(vertx, "openapi.yml")
                .flatMap(builder -> {
                    builder.setOptions(routerBuilderOptions);
                    operationHandlers.forEach(operationHandler -> {
                        final var operationName = operationHandler.operationName();
                        logger.fine(() -> "Registering " + operationName);
                        builder.operation(operationName).handler(operationHandler.handler());
                    });
                    builder.rootHandler(bodyHandler);
                    logger.fine(() -> "Registering Security handler Token");
                    builder.securityHandler("Token", authenticationHandler);
                    logger.fine(() -> "Creating httpServer");
                    return vertx.createHttpServer(httpServerOptions)
                            .requestHandler(builder.createRouter())
                            .listen()
                            .onSuccess(httpServer -> logger.info(() -> "Server listening on " + httpServer.actualPort()))
                            .onFailure(t -> logger.log(Level.SEVERE, "HttpServer failed to start listening.", t));
                }).onFailure(t -> logger.log(Level.SEVERE, "Failed to create a routerBuilder from openapi spec.", t));
    }

    @Component(modules = {VertxModule.class, UserModule.class, DatabaseModule.class, ConfigModule.class})
    @Singleton
    interface RealWorld {
        Main main();
    }
}
