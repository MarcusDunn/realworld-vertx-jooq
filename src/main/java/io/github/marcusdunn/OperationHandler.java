package io.github.marcusdunn;

import io.vertx.core.Handler;
import io.vertx.ext.web.RoutingContext;

public interface OperationHandler {
    String operationName();
    Handler<RoutingContext> handler();
}
