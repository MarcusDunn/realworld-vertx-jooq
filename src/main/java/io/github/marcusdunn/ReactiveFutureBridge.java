package io.github.marcusdunn;

import io.vertx.core.Future;
import io.vertx.core.Promise;
import java.util.Optional;
import java.util.concurrent.Flow;
import org.reactivestreams.Publisher;
import org.reactivestreams.Subscriber;
import org.reactivestreams.Subscription;

public class ReactiveFutureBridge {
    private ReactiveFutureBridge() {
    }

    public static <T> Future<Optional<T>> fetchOne(Publisher<T> publisher) {
        FutureSubscriber<T> s = new FutureSubscriber<>();
        publisher.subscribe(s);
        return s.getFuture();
    }

    private static class FutureSubscriber<T> implements Flow.Subscriber<T>, Subscriber<T> {
        private final Promise<Optional<T>> promise;

        public FutureSubscriber() {
            this.promise = Promise.promise();
        }

        public Future<Optional<T>> getFuture() {
            return promise.future();
        }

        @Override
        public void onSubscribe(Flow.Subscription subscription) {
            subscription.request(1);
        }

        @Override
        public void onSubscribe(Subscription s) {
            s.request(1);
        }

        @Override
        public void onNext(T t) {
            promise.complete(Optional.of(t));
        }

        @Override
        public void onError(Throwable t) {
            promise.fail(t);
        }

        @Override
        public void onComplete() {
            promise.tryComplete(Optional.empty());
        }
    }
}
