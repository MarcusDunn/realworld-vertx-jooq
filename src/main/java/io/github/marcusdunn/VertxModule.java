package io.github.marcusdunn;


import dagger.Module;
import dagger.Provides;
import io.vertx.core.Vertx;
import io.vertx.core.VertxOptions;
import io.vertx.core.http.HttpServerOptions;
import io.vertx.ext.auth.jwt.JWTAuth;
import io.vertx.ext.auth.jwt.JWTAuthOptions;
import io.vertx.ext.web.client.WebClient;
import io.vertx.ext.web.client.WebClientOptions;
import io.vertx.ext.web.handler.AuthenticationHandler;
import io.vertx.ext.web.handler.BodyHandler;
import io.vertx.ext.web.handler.JWTAuthHandler;
import io.vertx.ext.web.openapi.RouterBuilderOptions;
import org.apache.commons.configuration2.Configuration;

import javax.inject.Singleton;

@Module
public class VertxModule {
    private static String PORT_KEY = "PORT";
    private static int DEFAULT_PORT = 8080;
    @Provides
    @Singleton
    static Vertx vertx(VertxOptions vertxOptions) {
        return Vertx.vertx(vertxOptions);
    }

    @Provides
    static VertxOptions vertxOptions() {
        return new VertxOptions();
    }

    @Provides
    static AuthenticationHandler authenticationHandler(JWTAuth jwtAuth) {
        return JWTAuthHandler.create(jwtAuth);
    }

    @Provides
    static HttpServerOptions httpServerOptions(Configuration configuration) {
        return new HttpServerOptions()
                .setPort(configuration.getInt(PORT_KEY, DEFAULT_PORT));
    }

    @Provides
    static RouterBuilderOptions routerBuilderOptions() {
        return new RouterBuilderOptions();
    }

    @Provides
    static BodyHandler bodyHandler() {
        return BodyHandler.create();
    }

    @Provides
    static WebClient webClient(Vertx vertx, WebClientOptions webClientOptions) {
        return WebClient.create(vertx, webClientOptions);
    }

    @Provides
    static WebClientOptions webClientOptions() {
        return new WebClientOptions();
    }

    @Provides
    static JWTAuth jwtAuth(Vertx vertx, JWTAuthOptions jwtAuthOptions) {
        return JWTAuth.create(vertx, jwtAuthOptions);
    }

    @Provides
    static JWTAuthOptions jwtAuthOptions() {
        return new JWTAuthOptions();
    }
}
