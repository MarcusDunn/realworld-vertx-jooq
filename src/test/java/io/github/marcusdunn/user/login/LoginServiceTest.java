package io.github.marcusdunn.user.login;

import io.github.marcusdunn.AbstractDatabaseTest;
import io.github.marcusdunn.ReactiveFutureBridge;
import io.vertx.junit5.Timeout;
import io.vertx.junit5.VertxExtension;
import io.vertx.junit5.VertxTestContext;
import org.jooq.generated.tables.JUser;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import java.util.concurrent.TimeUnit;

import static io.github.marcusdunn.matcher.ApprovalMatcher.isApproved;
import static io.github.marcusdunn.matcher.EmptyValueMatcher.isEmpty;
import static io.github.marcusdunn.matcher.PresentValueMatcher.hasPresentValue;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.jooq.impl.DSL.value;

@ExtendWith(VertxExtension.class)
@Timeout(value = 2, timeUnit = TimeUnit.SECONDS)
class LoginServiceTest extends AbstractDatabaseTest {

    @Test
    void checkLoginWithNoMatchingUser(VertxTestContext testContext) {
        loginService
                .loginEmailPassword("hello", "world")
                .onComplete(testContext.succeeding(result -> testContext
                        .verify(() -> assertThat(result, isEmpty()))
                        .completeNow()));
    }

    @Test
    void checkLoginWithMatchingUser(VertxTestContext testContext) {
        String email = "marcus.dunn@example.com";
        String password = "password123";
        ReactiveFutureBridge.fetchOne(
                        dsl.insertInto(
                                        JUser.USER,
                                        JUser.USER.EMAIL,
                                        JUser.USER.PASSWORD
                                )
                                .values(
                                        value(email, JUser.USER.EMAIL),
                                        value(password, JUser.USER.PASSWORD)
                                )
                )
                .onComplete(testContext.succeeding(i -> loginService
                        .loginEmailPassword(email, password)
                        .onComplete(testContext.succeeding(result -> testContext
                                .verify(() -> assertThat(result, hasPresentValue(isApproved())))
                                .completeNow()
                        ))));
    }
}