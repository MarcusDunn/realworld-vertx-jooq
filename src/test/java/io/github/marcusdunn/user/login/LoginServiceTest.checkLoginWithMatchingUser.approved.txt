+----+-----------------------+-----------+
|  id|email                  |password   |
+----+-----------------------+-----------+
|   1|marcus.dunn@example.com|password123|
+----+-----------------------+-----------+
